libopae++/src/except.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright(c) 2017, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include <cerrno>
+#include <cstring>
+
+#include <safe_string/safe_string.h>
+#include <opae/utils.h>
+
+#include "opaec++/except.h"
+
+namespace opae
+{
+namespace fpga
+{
+namespace types
+{
+
+src_location::src_location(const char *file,
+                           const char *fn,
+                           int line) noexcept
+: file_(file)
+, fn_(fn)
+, line_(line)
+{}
+
+const char * src_location::file() const noexcept
+{
+    // return a pointer to the file name component.
+    const char *p = file_;
+
+    while (*p++) {}
+    while ((p > file_) &&
+           (*p != '\\') &&
+           (*p != '/'))
+        --p;
+    if (('\\' == *p) ||
+        ('/' == *p))
+        ++p;
+
+    return p;
+}
```

Comment:
Have you thought about using std::string operations to get the filename from the path? Like do a reverse find for `\\` and then call `substr` with that location?

Reply:
I was trying to process the exception without allocating any memory, in case the exception was caused by a low memory condition.

Reply:
Will placement new work?